### PR TITLE
Fix `torch._numpy.random`

### PR DESCRIPTION
**Pull request by ev-br**
Fix several issues with `torch._numpy.random` functions on eager

1. actually return scalars when `size is None`
2. fix dispatch with USE_NUMPY_STREAM
3. make tnp.random functions composable: make numpy functions receive numpy arguments, not `tnp.ndarray`s
4. fix random.shuffle for e.g. lists

The main need for this gymnastics is due to `np.random` functions returning an ndarray or python scalar depending on the `size` argument. We decided a while ago to replicate this behavior in `tnp.random` and not elsewhere where we always return 0D arrays instead.

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_numpy/random.py [hunk 1]
```diff
@@ -98,15 +116,33 @@ def normal(loc=0.0, scale=1.0, size=None):
...
+        _shuffle_tensor_inplace(x.tensor)
+    else:
+        # untyped code path. Apply the permutation in-place
+        # cf https://stackoverflow.com/questions/16501424/algorithm-to-apply-permutation-in-constant-memory-space/41472796#41472796
+        perm = torch.randperm(len(x)).tolist()
+        for i in range(len(x)):
+            idx = perm[i]
+            while idx < i:
+                idx = perm[idx]
+            x[i], x[idx] = x[idx], x[i]
```
Comment (lezcano) [inline torch/_numpy/random.py:139]: If the input is not a tensor let's just raise a NotImplementedError.

Comment (lezcano) [reply torch/_numpy/random.py:139]: After doing so, we can pretty much leave the implementation as it was modulo an early if

File: torch/_numpy/random.py [hunk 2]
```diff
@@ -64,7 +82,7 @@ def random_sample(size=None):
         size = ()
     dtype = _dtypes_impl.default_dtypes().float_dtype
     values = torch.empty(size, dtype=dtype).uniform_()
-    return array_or_scalar(values, return_scalar=size is None)
+    return array_or_scalar(values, return_scalar=size == ())
 
 
 @deco_stream
```
Comment (lezcano) [inline torch/_numpy/random.py:88]: Previous issue, but this decorator is not necessary, is it?

File: torch/_numpy/random.py [hunk 3]
```diff
@@ -82,7 +100,7 @@ def uniform(low=0.0, high=1.0, size=None):
         size = ()
     dtype = _dtypes_impl.default_dtypes().float_dtype
     values = torch.empty(size, dtype=dtype).uniform_(low, high)
-    return array_or_scalar(values, return_scalar=size is None)
+    return array_or_scalar(values, return_scalar=size == ())
 
 
 @deco_stream
```
Comment (lezcano) [inline torch/_numpy/random.py:106]: In this one we should probably just pass `return_scalar=False`.

File: torch/_numpy/random.py [hunk 4]
```diff
@@ -42,10 +42,28 @@ def inner(*args, **kwds):
...
+            )
+            kwds = {
+                key: val.tensor.numpy() if isinstance(val, ndarray) else val
+                for key, val in kwds.items()
+            }
+
+            value = f(*args, **kwds)
+
+            # `value` can be either _np.ndarray or python scalar (or None)
+            if value is not None and isinstance(value, _np.ndarray):
```
Comment (lezcano) [inline torch/_numpy/random.py:63]:
```suggestion
            if isinstance(value, _np.ndarray):
```

File: torch/_numpy/random.py [hunk 5]
```diff
@@ -42,10 +42,28 @@ def inner(*args, **kwds):
         if USE_NUMPY_RANDOM is False:
             return func(*args, **kwds)
         elif USE_NUMPY_RANDOM is True:
-            from numpy import random as nr
+            import numpy as _np
```
Comment (lezcano) [inline torch/_numpy/random.py:45]: Perhaps
```suggestion
            import numpy
```
as the `_np` thing is a bit odd.

File: torch/_numpy/random.py [hunk 6]
```diff
@@ -42,10 +42,28 @@ def inner(*args, **kwds):
         if USE_NUMPY_RANDOM is False:
             return func(*args, **kwds)
         elif USE_NUMPY_RANDOM is True:
```
Comment (lezcano) [inline torch/_numpy/random.py:44]: Now that we're at it, could you make this a flag in `_dynamo.config` near the other numpy flags that are there? I forgot to do so in the upstreaming PR

Comment (lezcano) [reply torch/_numpy/random.py:44]: Feel free to do it in this PR or in a follow up

**Review by lezcano — COMMENTED** [review]
Good catch. I left a few minor comments.

File: torch/_numpy/random.py (hunk 1 above)
Comment (ev-br) [reply torch/_numpy/random.py:139]: That would make our lives easier indeed, but it's kinda hard to explain. Previously, the random stream is either that of numpy or py torch, depending on the global flag. 

With a graph break, the stream depends on the argument type.

File: torch/_numpy/random.py (hunk 2 above)
Comment (ev-br) [reply torch/_numpy/random.py:88]: On eager it is, why?

File: torch/_numpy/random.py (hunk 3 above)
Comment (ev-br) [reply torch/_numpy/random.py:106]: Also why? In numpy `uniform(0, 1) ` does returns python float

File: torch/_numpy/random.py (hunk 1 above)
Comment (lezcano) [reply torch/_numpy/random.py:139]: If we graphbreak the stream is always going to be that of NumPy, so that's consistent all across

Comment (ev-br) [reply torch/_numpy/random.py:139]: Under dynamo, yes. Do not care about eager at all?

Comment (lezcano) [reply torch/_numpy/random.py:139]: Not to this extent. Any place were we break via a `NotImplementedError` is going to differ between eager and compiled, but that's alright.

File: torch/_numpy/random.py (hunk 3 above)
Comment (lezcano) [reply torch/_numpy/random.py:106]: I commented here referring to the function below, as I couldnt' comment in the lines below

File: torch/_numpy/random.py (hunk 2 above)
Comment (lezcano) [reply torch/_numpy/random.py:88]: nb. This comment refers to `rand`.

File: test/torch_np/test_random.py [hunk 7]
```diff
@@ -2,54 +2,127 @@
...
+        ("normal", ()),
+        ("rand", ()),
+        ("randint", (0, 5)),
+        ("randn", ()),
+        ("random", ()),
+        ("random_sample", ()),
+        ("sample", ()),
+        ("uniform", ()),
+    ],
+)
```
Comment (lezcano) [inline test/torch_np/test_random.py:39]: You can simply put the `tnp.random.*` functions here. Then you can bind `partial(tnp.random.randint, 0, 5)` and remove the `arg` parameter. If you do this, the `test_array` can be written much clearer as
```python
def test_array(self, fn, use_numpy):
    with control_stream(use_numpy):
        if name in ("rand", "randn"):
            r = fn(10)
        else:
            r = fn(size=10)
    assert isinstance(r, tnp.ndarray)
```
and in the scalar one you can simply call `fn()`.

File: test/torch_np/test_random.py [hunk 8]
```diff
@@ -2,54 +2,127 @@
...
+
+
+@pytest.mark.parametrize("use_numpy", [True, False])
+def test_choice(use_numpy):
+    kwds = dict(size=3, replace=False, p=[0.1, 0, 0.3, 0.6, 0])
+    with control_stream(use_numpy):
+        tnp.random.seed(12345)
+        x = tnp.random.choice(5, **kwds)
+        x_1 = tnp.random.choice(tnp.arange(5), **kwds)
+        assert_equal(x, x_1)
```
Comment (lezcano) [inline test/torch_np/test_random.py:101]: Why do two consecutive calls return the same?

File: torch/_numpy/random.py [hunk 9]
```diff
@@ -98,15 +116,26 @@ def normal(loc=0.0, scale=1.0, size=None):
...
+    return array_or_scalar(values, return_scalar=size == ())
 
 
 @deco_stream
-@normalizer
 def shuffle(x: ArrayLike):
-    perm = torch.randperm(x.shape[0])
-    xp = x[perm]
-    x.copy_(xp)
+    # no @normalizer because we need to shuffle e.g. lists in-place
```
Comment (lezcano) [inline torch/_numpy/random.py:130]: Because we don't want to cast lists to tensors*
We should also remove the `ArrayLike` decorator.

File: torch/_numpy/random.py [hunk 10]
```diff
@@ -98,15 +116,26 @@ def normal(loc=0.0, scale=1.0, size=None):
...
+
+    if isinstance(x, torch.Tensor):
+        _shuffle_tensor_inplace(x)
+    elif isinstance(x, ndarray):
+        _shuffle_tensor_inplace(x.tensor)
+    else:
+        raise NotImplementedError("We do not random.shuffle lists in-place")
+
+
+def _shuffle_tensor_inplace(t):
```
Comment (lezcano) [inline torch/_numpy/random.py:141]: nit. we can inline again this one.

**Comment by ev-br** [comment]
I believe I addressed all @lezcano's review comments.

File: torch/_dynamo/config.py [hunk 11]
```diff
@@ -266,6 +266,9 @@
 numpy_default_complex = "complex128"
 numpy_default_int = "int64"
 
+# use numpy random variates if True, pytorch otherwise
```
Comment (lezcano) [inline torch/_dynamo/config.py:269]: variates may be a bit too mathematical?
```suggestion
# use numpy's PRNG if True, pytorch otherwise
```

File: test/torch_np/test_random.py [hunk 12]
```diff
@@ -2,52 +2,127 @@
...
-    x = tnp.random.uniform(0, 1, size=11)
-
-    # check that the stream is identical to numpy's
-    import numpy as _np
-
-    _np.random.seed(12345)
-    x_np = _np.random.uniform(0, 1, size=11)
+@contextmanager
+def control_stream(use_numpy=False):
+    import torch._dynamo.config as config
```
Comment (lezcano) [inline test/torch_np/test_random.py:16]: Make it global.

File: torch/_numpy/random.py [hunk 13]
```diff
@@ -29,25 +29,48 @@
...
+                value = ndarray(torch.as_tensor(value))
 
-            f = getattr(nr, func.__name__)
-            return f(*args, **kwds)
+            return value
         else:
-            raise ValueError(f"USE_NUMPY_RANDOM={USE_NUMPY_RANDOM} not understood.")
+            raise ValueError(
+                f"config.use_numpy_random_stream = {use_numpy_random()} not understood."
+            )
```
Comment (lezcano) [inline torch/_numpy/random.py:73]: Not needed anymore. We expect people to respect the python types. Same above,  `is True / is False` is not necessary.

**Review by lezcano — COMMENTED** [review]
Mostly nits. The only one that I did not quite understand is why `test_choice` passes.

File: test/torch_np/test_random.py (hunk 8 above)
Comment (ev-br) [reply test/torch_np/test_random.py:101]: This is numpy: https://numpy.org/doc/stable/reference/random/generated/numpy.random.RandomState.choice.html

```
a : 1-D array-like or int
If an ndarray, a random sample is generated from its elements. 
If an int, the random sample is generated as if it were np.arange(a)
```

(I know)

Comment (lezcano) [reply test/torch_np/test_random.py:99]: Oh wow, this is not what I guess this function was doing lol

**Review by lezcano — APPROVED** [review]

**Comment by ev-br** [comment]
Comments addressed

**Comment by lezcano** [comment]
Did you check that the failures on huggingface are not from this PR?

**Comment by lezcano** [comment]
Yep, according to https://hud2.pytorch.org/, those failures were there from before (from the SEV). Cool.
